File: data-prep/dataprep-model/src/main/scala/de/hpi/isg/dataprep/spark/ChangePropertyDataTypeImpl.scala
package de.hpi.isg.dataprep.spark

import de.hpi.isg.dataprep.Consequences
import de.hpi.isg.dataprep.SparkPreparators.spark
import de.hpi.isg.dataprep.preparators.ChangePropertyDataType.PropertyType
import org.apache.spark.sql.catalyst.encoders.RowEncoder
import org.apache.spark.sql.types.{IntegerType, StructField, StructType}
import org.apache.spark.sql.{DataFrame, Row, SparkSession}
import org.apache.spark.util.CollectionAccumulator

import scala.util.{Failure, Success, Try}
import spark.implicits._

/**
  * @author Lan Jiang
  * @since 2018/8/7
  */
object ChangePropertyDataTypeImpl {

    def changePropertyDataType(dataFrame: DataFrame, columnName: String, sourceDataType: PropertyType, targetDataType: PropertyType): Consequences = {
        new Consequences(new CollectionAccumulator[(Any, Throwable)])
    }

    def changePropertyDataType(dataFrame: DataFrame, fieldName: String, targetDataType: PropertyType): Consequences = {
        val errorsAccumulator = new CollectionAccumulator[(Any, Throwable)]
        dataFrame.sparkSession.sparkContext.register(errorsAccumulator, "test")

        val transformed = dataFrame.rdd.flatMap(row => {
            val fe = Try {
                targetDataType match {
                    case PropertyType.INTEGER => row.getAs[String](fieldName).toInt
                    case PropertyType.STRING => row.getAs[String](fieldName).toString
                }
            }
            val trial = fe match {
                case Failure(t) => {
                    errorsAccumulator.add(row.getAs[String](fieldName), t)
                    fe
                }
                case valid: Success[t] => {
                    valid
                }
            }
            trial.toOption
        }).count() // count to update the accumulator
        System.out.println(errorsAccumulator.value)

        new Consequences(errorsAccumulator)
    }
}

File: data-prep/dataprep-preparators/src/main/java/de/hpi/isg/dataprep/implementation/ChangePropertyDataTypeImpl.java
package de.hpi.isg.dataprep.implementation;

import de.hpi.isg.dataprep.Consequences;
import de.hpi.isg.dataprep.model.error.PreparationError;
import de.hpi.isg.dataprep.model.error.RecordError;
import de.hpi.isg.dataprep.model.target.preparator.Preparator;
import de.hpi.isg.dataprep.model.target.preparator.PreparatorImpl;
import de.hpi.isg.dataprep.preparators.ChangePropertyDataType;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.util.CollectionAccumulator;

/**
 * The delegation class for {@link ChangePropertyDataType}. Maybe not needed.
 *
 * @author Lan Jiang
 * @since 2018/8/9
 */
abstract public class ChangePropertyDataTypeImpl extends PreparatorImpl {

    abstract protected Consequences executeLogic(ChangePropertyDataType preparator,
                                                 Dataset<Row> dataFrame,
                                                 CollectionAccumulator<PreparationError> errorAccumulator);

    @Override
    protected final Consequences executePreparator(Preparator preparator, Dataset<Row> dataFrame) throws Exception {
        ChangePropertyDataType preparator_ = this.getPreparatorInstance(preparator, ChangePropertyDataType.class);
        CollectionAccumulator<PreparationError> errorAccumulator = this.createErrorAccumulator(preparator_, dataFrame);

        return executeLogic(preparator_, dataFrame, errorAccumulator);
    }

    @Override
    protected final CollectionAccumulator<PreparationError> createErrorAccumulator(Preparator preparator, Dataset<Row> dataFrame) {
        CollectionAccumulator<PreparationError> errorAccumulator = new CollectionAccumulator();
        dataFrame.sparkSession().sparkContext().register(errorAccumulator,
                String.format("%s error accumulator.", ChangePropertyDataType.class.getSimpleName()));
        return errorAccumulator;
    }
}